sympy/core/operations.py

```diff
@@ -504,3 +505,118 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(OpDispatcher):
+    """
+    Handler dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    If a new class is meant to return the result other than `Add` or `Mul` when operated,
+    it should have its handler defined as class method. Also, if the arguments have the same
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if issubclass(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    Notes
+    =====
+
+    This approach is experimental, and can be replaced or deleted in the future.
+    See https://github.com/sympy/sympy/pull/19463.
+
+    """
+
+    def register(self, *args, **kwargs):
+        raise NotImplementedError("To avoid confusion, use 'register_priority' method instead.")
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
+        """
+        def _(func):
+            self.add((cls1, cls2), func)
+            return func
+        return _
+
```

Comment:
With my suggestion below, this becomes
```suggestion
    def register_priority(self, cls1, cls2):
        """
        Register the priority between two types. Registered function must
        return the type which has the handler that is to be used.
        """
        return self._type_dispatcher.register(cls1, cls2)
```
The fact that you're having to remove the inherited `Dispatcher` methods is a big clue that subclassing `Dispatcher` doesn't make a whole lot of sense.